IllegalReferenceCountException logged with enabled compression

Hi

**Describe the bug**
I'm seeing this in 1.11.1.Final after enabling compression via `quarkus.http.enable-compression`:
```
2021-01-29 07:19:42,608 WARN  [io.net.cha.ChannelOutboundBuffer] (vert.x-eventloop-thread-4) Failed to mark a promise as success because it has failed already: DefaultChannelPromise@3d236c66(failure: io.netty.handler.codec.EncoderException: io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1), unnotified cause:: io.netty.handler.codec.EncoderException: io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1
        at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:104)
        at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:717)
        at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:709)
        at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:792)
        at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:702)
        at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:697)
        at io.netty.handler.codec.MessageToMessageEncoder.writePromiseCombiner(MessageToMessageEncoder.java:137)
        at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:117)
        at io.netty.handler.codec.MessageToMessageCodec.write(MessageToMessageCodec.java:116)
        at io.vertx.core.http.impl.HttpChunkContentCompressor.write(HttpChunkContentCompressor.java:38)
        at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:717)
        at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:709)
        at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:792)
        at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:702)
        at io.netty.handler.stream.ChunkedWriteHandler.doFlush(ChunkedWriteHandler.java:300)
        at io.netty.handler.stream.ChunkedWriteHandler.flush(ChunkedWriteHandler.java:132)
        at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:750)
        at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:742)
        at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:728)
        at io.netty.channel.ChannelDuplexHandler.flush(ChannelDuplexHandler.java:127)
        at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:750)
        at io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:765)
        at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:790)
        at io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:758)
        at io.vertx.core.net.impl.ConnectionBase.write(ConnectionBase.java:124)
        at io.vertx.core.net.impl.ConnectionBase.lambda$queueForWrite$2(ConnectionBase.java:215)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1
        at io.netty.util.internal.ReferenceCountUpdater.toLiveRealRefCnt(ReferenceCountUpdater.java:74)
        at io.netty.util.internal.ReferenceCountUpdater.release(ReferenceCountUpdater.java:138)
        at io.netty.buffer.AbstractReferenceCountedByteBuf.release(AbstractReferenceCountedByteBuf.java:100)
        at io.vertx.core.http.impl.AssembledHttpResponse.release(AssembledHttpResponse.java:159)
        at io.netty.util.ReferenceCountUtil.release(ReferenceCountUtil.java:88)
        at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:91)
        ... 32 more
```


**Expected behavior**
There should be no exception logged as the content is transfered without any problem.

**To Reproduce**

Steps to reproduce the behavior:
1. Enable compression
2. Access a (reactive [Mutiny]?) endpoint

I'll try to create a reproducer if needed be.

**Configuration**
```properties
quarkus:
  http:
    auth:
      proactive: true
    cors:
      ~: true
      origins: http://localhost:3000
      methods: DELETE, GET, POST, PUT, OPTIONS
    root-path: /
    limits:
      max-body-size: 20480K
    enable-compression: true
  resteasy:
    gzip:
      enabled: true
      max-input: 20480K
```

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: `Linux voyager 5.10.11-arch1-1 #1 SMP PREEMPT Wed, 27 Jan 2021 13:53:16 +0000 x86_64 GNU/Linux`
 - Output of `java -version`: 
 ```
openjdk version "11.0.10" 2021-01-19
OpenJDK Runtime Environment (build 11.0.10+8)
OpenJDK 64-Bit Server VM (build 11.0.10+8, mixed mode)
```
 - GraalVM version (if different from Java): 
 - Quarkus version or git rev: 1.11.1.Final
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): 
```
❯ ./gradlew --version

------------------------------------------------------------
Gradle 6.8.1
------------------------------------------------------------

Build time:   2021-01-22 13:20:08 UTC
Revision:     31f14a87d93945024ab7a78de84102a3400fa5b2

Kotlin:       1.4.20
Groovy:       2.5.12
Ant:          Apache Ant(TM) version 1.10.9 compiled on September 27 2020
JVM:          11.0.10 (Oracle Corporation 11.0.10+8)
OS:           Linux 5.10.11-arch1-1 amd64
```

**Additional context**
* Adding "quarkus-undertow" dependency like stated in #9194 does _not_ remove the exception.
* Disabling compression removes the exception.


/edit: fixed the statement about the tried workaround – it's not working.

--- a/bom/application/pom.xml
+++ b/bom/application/pom.xml
@@ -53,7 +53,7 @@
         <smallrye-context-propagation.version>1.2.0</smallrye-context-propagation.version>
         <smallrye-reactive-streams-operators.version>1.0.13</smallrye-reactive-streams-operators.version>
         <smallrye-reactive-types-converter.version>2.6.0</smallrye-reactive-types-converter.version>
-        <smallrye-mutiny-vertx-binding.version>2.12.0</smallrye-mutiny-vertx-binding.version>
+        <smallrye-mutiny-vertx-binding.version>2.13.0</smallrye-mutiny-vertx-binding.version>
         <smallrye-reactive-messaging.version>3.9.1</smallrye-reactive-messaging.version>
         <jakarta.activation.version>1.2.1</jakarta.activation.version>
         <jakarta.annotation-api.version>1.3.5</jakarta.annotation-api.version>
@@ -109,7 +109,7 @@
         <wildfly-elytron.version>1.16.1.Final</wildfly-elytron.version>
         <jboss-modules.version>1.8.7.Final</jboss-modules.version>
         <jboss-threads.version>3.4.2.Final</jboss-threads.version>
-        <vertx.version>4.1.2</vertx.version>
+        <vertx.version>4.1.3</vertx.version>
         <httpclient.version>4.5.13</httpclient.version>
         <httpcore.version>4.4.14</httpcore.version>
         <httpasync.version>4.1.4</httpasync.version>
@@ -131,7 +131,7 @@
         <infinispan.version>12.1.7.Final</infinispan.version>
         <infinispan.protostream.version>4.4.1.Final</infinispan.protostream.version>
         <caffeine.version>2.9.2</caffeine.version>
-        <netty.version>4.1.65.Final</netty.version>
+        <netty.version>4.1.67.Final</netty.version>
         <reactive-streams.version>1.0.3</reactive-streams.version>
         <jboss-logging.version>3.4.2.Final</jboss-logging.version>
         <mutiny.version>1.0.0</mutiny.version>

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/LargerThanDefaultFormAttributeMultipartFormInputTest.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/LargerThanDefaultFormAttributeMultipartFormInputTest.java
@@ -39,7 +39,7 @@ public class LargerThanDefaultFormAttributeMultipartFormInputTest {
                     return ShrinkWrap.create(JavaArchive.class)
                             .addClasses(Resource.class, Data.class)
                             .addAsResource(new StringAsset(
-                                    "quarkus.http.limits.max-form-attribute-size=4K"),
+                                    "quarkus.http.limits.max-form-attribute-size=120K"),
                                     "application.properties");
                 }
             });
@@ -49,6 +49,12 @@ public class LargerThanDefaultFormAttributeMultipartFormInputTest {
     @Test
     public void test() throws IOException {
         String fileContents = new String(Files.readAllBytes(FILE.toPath()), StandardCharsets.UTF_8);
+        StringBuilder sb = new StringBuilder();
+        for (int i = 0; i < 10; ++i) {
+            sb.append(fileContents);
+        }
+        fileContents = sb.toString();
+
         Assertions.assertTrue(fileContents.length() > HttpServerOptions.DEFAULT_MAX_FORM_ATTRIBUTE_SIZE);
         given()
                 .multiPart("text", fileContents)

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/TooLargeFormAttributeMultipartFormInputTest.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/TooLargeFormAttributeMultipartFormInputTest.java
@@ -66,9 +66,13 @@ public class TooLargeFormAttributeMultipartFormInputTest extends AbstractMultipa
 
     @Test
     public void test() throws IOException {
-        String formAttrSourceFileContents = new String(Files.readAllBytes(FORM_ATTR_SOURCE_FILE.toPath()),
-                StandardCharsets.UTF_8);
-        Assertions.assertTrue(formAttrSourceFileContents.length() > HttpServerOptions.DEFAULT_MAX_FORM_ATTRIBUTE_SIZE);
+        String fileContents = new String(Files.readAllBytes(FORM_ATTR_SOURCE_FILE.toPath()), StandardCharsets.UTF_8);
+        StringBuilder sb = new StringBuilder();
+        for (int i = 0; i < 10; ++i) {
+            sb.append(fileContents);
+        }
+        fileContents = sb.toString();
+        Assertions.assertTrue(fileContents.length() > HttpServerOptions.DEFAULT_MAX_FORM_ATTRIBUTE_SIZE);
         given()
                 .multiPart("active", "true")
                 .multiPart("num", "25")
@@ -76,7 +80,7 @@ public class TooLargeFormAttributeMultipartFormInputTest extends AbstractMultipa
                 .multiPart("htmlFile", HTML_FILE, "text/html")
                 .multiPart("xmlFile", XML_FILE, "text/xml")
                 .multiPart("txtFile", TXT_FILE, "text/plain")
-                .multiPart("name", formAttrSourceFileContents)
+                .multiPart("name", fileContents)
                 .accept("text/plain")
                 .when()
                 .post("/test")

--- a/extensions/vertx-http/runtime/pom.xml
+++ b/extensions/vertx-http/runtime/pom.xml
@@ -53,6 +53,11 @@
                 </exclusion>
             </exclusions>
         </dependency>
+        <dependency>
+            <groupId>org.graalvm.nativeimage</groupId>
+            <artifactId>svm</artifactId>
+            <scope>provided</scope>
+        </dependency>
         <!-- Test dependencies -->
         <dependency>
             <groupId>org.junit.jupiter</groupId>

--- a/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/graal/HttpContentCompressorSubstitutions.java
+++ b/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/graal/HttpContentCompressorSubstitutions.java
@@ -1,0 +1,115 @@
+package io.quarkus.vertx.http.runtime.graal;
+
+import static io.netty.handler.codec.http.HttpHeaderValues.DEFLATE;
+import static io.netty.handler.codec.http.HttpHeaderValues.GZIP;
+import static io.netty.handler.codec.http.HttpHeaderValues.X_DEFLATE;
+import static io.netty.handler.codec.http.HttpHeaderValues.X_GZIP;
+
+import java.util.function.BooleanSupplier;
+
+import com.oracle.svm.core.annotate.Alias;
+import com.oracle.svm.core.annotate.Substitute;
+import com.oracle.svm.core.annotate.TargetClass;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.embedded.EmbeddedChannel;
+import io.netty.handler.codec.compression.ZlibWrapper;
+import io.netty.handler.codec.http2.CompressorHttp2ConnectionEncoder;
+import io.netty.handler.codec.http2.Http2Exception;
+
+public class HttpContentCompressorSubstitutions {
+
+    @TargetClass(className = "io.netty.handler.codec.compression.ZstdEncoder", onlyWith = IsZstdAbsent.class)
+    public static final class ZstdEncoderFactorySubstitution {
+
+        @Substitute
+        protected ByteBuf allocateBuffer(ChannelHandlerContext ctx, ByteBuf msg, boolean preferDirect) throws Exception {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        protected void encode(ChannelHandlerContext ctx, ByteBuf in, ByteBuf out) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        public void flush(final ChannelHandlerContext ctx) {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    @TargetClass(className = "io.netty.handler.codec.compression.BrotliEncoder", onlyWith = IsBrotliAbsent.class)
+    public static final class BrEncoderFactorySubstitution {
+
+        @Substitute
+        protected ByteBuf allocateBuffer(ChannelHandlerContext ctx, ByteBuf msg, boolean preferDirect) throws Exception {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        protected void encode(ChannelHandlerContext ctx, ByteBuf in, ByteBuf out) {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    @TargetClass(CompressorHttp2ConnectionEncoder.class)
+    public static final class CompressorHttp2ConnectionSubstitute {
+
+        @Substitute
+        protected EmbeddedChannel newContentCompressor(ChannelHandlerContext ctx, CharSequence contentEncoding)
+                throws Http2Exception {
+            if (GZIP.contentEqualsIgnoreCase(contentEncoding) || X_GZIP.contentEqualsIgnoreCase(contentEncoding)) {
+                return newCompressionChannel(ctx, ZlibWrapper.GZIP);
+            }
+            if (DEFLATE.contentEqualsIgnoreCase(contentEncoding) || X_DEFLATE.contentEqualsIgnoreCase(contentEncoding)) {
+                return newCompressionChannel(ctx, ZlibWrapper.ZLIB);
+            }
+            // 'identity' or unsupported
+            return null;
+        }
+
+        @Alias
+        private EmbeddedChannel newCompressionChannel(final ChannelHandlerContext ctx, ZlibWrapper wrapper) {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    public static class IsZstdAbsent implements BooleanSupplier {
+
+        private boolean zstdAbsent;
+
+        public IsZstdAbsent() {
+            try {
+                Class.forName("com.github.luben.zstd.Zstd");
+                zstdAbsent = false;
+            } catch (ClassNotFoundException e) {
+                zstdAbsent = true;
+            }
+        }
+
+        @Override
+        public boolean getAsBoolean() {
+            return zstdAbsent;
+        }
+    }
+
+    public static class IsBrotliAbsent implements BooleanSupplier {
+
+        private boolean brotliAbsent;
+
+        public IsBrotliAbsent() {
+            try {
+                Class.forName("com.aayushatharva.brotli4j.encoder.Encoder");
+                brotliAbsent = false;
+            } catch (ClassNotFoundException e) {
+                brotliAbsent = true;
+            }
+        }
+
+        @Override
+        public boolean getAsBoolean() {
+            return brotliAbsent;
+        }
+    }
+}

--- a/independent-projects/resteasy-reactive/pom.xml
+++ b/independent-projects/resteasy-reactive/pom.xml
@@ -52,7 +52,7 @@
         <jakarta.json.version>1.1.6</jakarta.json.version>
         <mutiny.version>1.0.0</mutiny.version>
         <smallrye-common.version>1.6.0</smallrye-common.version>
-        <vertx.version>4.1.2</vertx.version>
+        <vertx.version>4.1.3</vertx.version>
         <rest-assured.version>4.4.0</rest-assured.version>
         <commons-logging-jboss-logging.version>1.0.0.Final</commons-logging-jboss-logging.version>
         <jboss-jaxb-api_2.3_spec.version>2.0.0.Final</jboss-jaxb-api_2.3_spec.version>

